conversation_simulator/src/StateMachine.js

```diff
@@ -142,67 +142,74 @@ export default class {
+        let submittedWords = this.splitString(solutionPhrase)
         let solutionWords = this.splitString(possibleSolution)
-        if (possibleSolution !== 'default' && this.checkSolution(submittedWords, solutionWords)) {
+        if (possibleSolution !== 'default' && this.compareSolution(submittedWords, solutionWords)) {
           solution = possibleSolution
           break
         }
       }
     }
-    let success = solution !== undefined ? 'True' : 'False'
-    // Apply score
-    dtml.scorePhrase(normalizedPhrase, success, result => {
-      if (result) {
-        this.scoreSolution(solution, result, noPoints)
-      } else {
-        let score = this.currentState.Solutions[solution].Score || this.currentState.Solutions[solution].scoreadjustment
-        if (this.isNumber(score)) {
-          this.scoreSolution(solution, score, noPoints)
+    if (solution === null) {
+      // Return default if it exists
+      if (this.currentState.Solutions['default']) return 'default'
+    }
+    return solution
+  }
+
+  // Returns the score of a solution
+  scoreSolution (solution, isCorrect, source) {
+    let _this = this
+    return new Promise(function (resolve, reject) {
+      dtml.scorePhrase(solution, isCorrect, result => {
+        if (result) {
+          resolve(result)
+        } else {
+          let score = _this.currentState.Solutions[solution].Score || _this.currentState.Solutions[solution].scoreadjustment
+          if (_this.isNumber(score)) {
+            resolve(score)
+          } else {
+            reject(new Error('Could not find a score'))
+          }
         }
-      }
-    }, source, this.getCurrentStateName())
+      }, source, _this.getCurrentStateName())
+    })
   }
 
-  scoreSolution (solution, score, noPoints) {
-    if (solution !== undefined) {
-      let nextState = this.currentState.Solutions[solution].Next
-      this.submitSolutionResult = true
-      if (score > 0 && !noPoints && !this.userContext.hasSolutionBeenUsed(this.getCurrentStateName(), solution)) {
-        this.score += score
-      }
-      // Save state/answer so it can't be used again for points
-      this.userContext.markSolution(this.getCurrentStateName(), solution)
-      // Go to next state
-      this.setCurrentState(
-        nextState,
-        this.stateData.States[nextState]
-      )
-    } else {
-      // If there is a default next state then it doesn't count as a failure
-      if (this.currentState.Solutions['default'].Next !== null) {
-        this.submitSolutionResult = true
-        let nextState = this.currentState.Solutions['default'].Next
-        this.setCurrentState(
-          nextState,
-          this.stateData.States[nextState]
-        )
-      } else {
-        this.submitSolutionResult = false
-        this.score -= 10
-      }
-    }
+  // Adds the score to the player's total score
+  applyScore (score) {
+    this.score += score
+  }
+
+  markSolution (stateName, solution) {
```

Comment:
Why do we need this proxy functions? Why not use useContext directly?

Reply:
I added the proxy function because I was calling it from the main game file. I thought it would be a bad practice to call the state machine's helper files directly from an external file.

Reply:
I see, I need to look deeper to understand how it is used, but at the high level we should use componsition as such that state machine is initialized with user context. I see you point but current approach couples usercontext and state machine. I would go somethig like stateMachine.GetUserConext().hasSolutionBeenUsed(stateName, solution);
In that case State Machie is not aware of userConext inteface exposed outside and doesnt need to change if user Context object changes

Reply:
Ok understood! That's how it's set up now.